### PR TITLE
Ported the test infrastructure to Tasty

### PR DESCRIPTION
**Pull request by gpyh**
This solves #2927. It should not introduce regression in the test suite.

**The text below is inaccurate and left for posterity. You should check the commit messages and the `test/README.md` file for up-to-date information.**

## Summary of the changes

The functional changes are the following:

- The output is pretty, even in CI logs (colors, individual timings, proper indentation) ; Tasty.Golden does the heavy lifting, i.e. comparing outputs and measuring time spent.
- No more reliance on `make` to call the script on selected tests. `runtest` now does a full sweep.
You can use the pattern feature from Tasty if you want to run only some selection of tests (see below)
- Tests will run in parallel in Travis CI. Furthermore, contrary to make, Tasty will make sure that tests results are always displayed in the same order. 
- Tests are run using `cabal test`, so we don't need to "guess" the path to the idris executable anymore. Cabal takes care of temporary adding the right directory to `PATH`.
- Tests are organised in families, reflecting directories' prefixes.
- New families must be added manually in `runtest`. It is not needed to
manually add new tests that belong to an existing family as `runtest` scans the directory, looking for directories having the right prefix.
- If a test crashes, it is reported as a failure instead of halting the whole
process.
- Previous configuration options have been removed. We still have feature parity though.
- A custom option has been implemented to run the test suite with the node code generator.
Incompatible tests must be given in the source code of `runtest`. This will be changed as to decouple code and data.

### Basics

Do `make test_c` or `make test_js`. It will run the tests in parallel and through `cabal test`.
You can also manually do `cabal test` or `stack test`

### Passing arguments

You can pass arguments to the test program. For example, to pass the `--help` argument:

```
# Via make
make TEST-ARGS="--help" test_c
# Via cabal
cabal test --test-options="--help"
# Via stack
stack test --test-arguments="--help"
```

Try it to learn more about the other arguments you can pass. Two are of particular interest: `--node` to test against the JS FFI and the Node code generator and `--pattern <regex>` to only run the test that match the provided `<regex>`.

### Specifying the number of parallel jobs

With make, the test suite runs in parallel by default. You can specify the number of threads with `TEST-JOBS`. For stack and cabal, you need to explicitly enable parallelism as you would do with any other GHC-compiled executable. Examples:

```
# Two test jobs via make
make TEST-JOBS=2 test_c
# Two test jobs via cabal
cabal test --test-options="+RTS -N2 -RTS"
# Two test jobs via stack
stack test --test-arguments="+RTS -N2 -RTS"
```

### Running only previously failed tests

Because of the `--rerun-update` option, `make test` will create a `.tasty-rerun-log` in the root directory of the project. Each time the test suite is run, the file will be written with the result of the tests. The next time you do `make test`, you can use the `rerun-filter` to, for example, only run previously failed tests. Pass "--help" to output the valid values.

### Other make rules

The make rule `test_times` has been removed as the tests are now always timed.
`test_clean` will rougly remove the tests outputs on the file system (some intermediate files will stay)
and `test_update` replaces the golden files with the current test outputs.

This is just preliminary work. I hope it is good enough to be merged.
I am waiting on your feedback on #3228 to move forward with the changes.

Also, I took the liberty to add folding in the Travis log. I used it **a lot** when coding and I grew tired of scrolling through the setup phase. I suppose it can stay that way.

### CODE REVIEW

File: stack-shell.nix [hunk 1]
```diff
@@ -9,6 +9,7 @@ let
     zlib
     gmp
     pkgconfig
+    ncurses
   ] ++ lib.optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
```
Comment (melted) [inline stack-shell.nix]: Is this related? Seems weird.

Comment (gpyh) [inline stack-shell.nix]: No it is not. It was just a missing dependency on the Nix script that prevented me to compile Idris. It kinda slipped out.

**Comment by gpyh** [comment]
The Travis job failed because it doesn't use `cabal` to build the test script.
This proposal requires it in order to cleanly fetch the path of the idris executable.

I will change this so the test script will directly link to the idris library instead of calling the executable.

It will have the added benefit to prevent unneeded compilations in some cases.

**Comment by ahmadsalim** [comment]
@gpyh Why did you close this PR?

**Comment by gpyh** [comment]
I closed this PR because it would never pass Travis.

Previously, the test script did some ugly string mangling to guess the path of the idris executable.
To clean this up, I imported a special module generated by cabal that would directly give access to such information. The problem is that Travis doesn't use cabal to run the tests but the Makefile under `test/`.

As such, I can see two solutions to this problem: the short term one would be to replace the current make rules with one calling cabal. As I don't know the CI infrastructure I don't understand the consequences of such a change. The long term one would be to forgo the use of the executable all together. As the idris executable's code is just a wrapper around a library, the run script could directly call the library.

This has the added benefit that, in the potential future architecture, we would pass CLI arguments to idris with the `Opt` datatype in `Idris.AbsSyntaxTree` instead of passing them as strings, which is error-prone.

**Comment by ahmadsalim** [comment]
@gpyh I think you should reopen it and then people here can examine the Travis issues with you :). I don't think it is as big a deal as you make it, and we don't want your work going to waste.

**Comment by ahmadsalim** [comment]
@gpyh You can see the Travis configuration here: https://github.com/idris-lang/Idris-dev/blob/master/.travis.yml . I think the easiest would be to see whether you could replace the call to the test script accordingly.

**Comment by ahmadsalim** [comment]
@gpyh What is the status of this PR?

**Comment by gpyh** [comment]
On hold, and will probably stay that way for the coming week as I'm pretty busy at work.

I still can't have both cabal running the tests and a working CI environment, despite numerous tries on both fronts. Once this is solved we'll be able to merge this.

**Comment by ahmadsalim** [comment]
@gpyh So it seems that you runtest.hs file needs `{-# LANGUAGE DeriveDataTypeable #-}` at the top

**Comment by gpyh** [comment]
I went at it again.
I added DeriveDataTypeable for versions of GHC prior to 7.10 and it (partly) solved the problem.
Now, it seems that tests running with cabal can't find the idris libraries. I poked around setting the `TARGET` and `IDRIS_LIBRARY_PATH` at various points but it did not solve the problem. I really need the help of someone more knowledgeable with cabal and idris path resolutions.

Unrelated: as requested, failing tests display the test output and the expected output directly in the log.

**Comment by gpyh** [comment]
The bug seems to be on cabal side and a workaround has been found. Will update the PR shortly.
Thanks @jfdm for helping me get my head out of the sand.

**Comment by gpyh** [comment]
This is "ready" to be merged, meaning that the port is done.

However, it now causes two tests to fail for the Node backend: `primitives003` and `totality009`.
I don't understand what makes them so special.

(Btw, top post has been edited with extensive info)

**Comment by gpyh** [comment]
@ahmadsalim This is ready to be merged, provided you're satisfied with the work.

The top post has been updated with instructions.

I intend to keep working on this. #3231 will track the evolution.

**Comment by ahmadsalim** [comment]
@gpyh Thanks for sticking through the PR! It seems like a very useful contribution.

 I think for a PR this big it might be a good idea that we get a review from someone previously involved in the testing infrastructure like @edwinb, @jfdm or @melted , who can take a better look and merge it if acceptable when they have time. If no one have looked at it for some time, I will take a final look myself and see if I can merge it.

**Comment by ahmadsalim** [comment]
I have taken a quick look and it seems that everything looks great so far (and the tests run a bit faster). So :thumbs: from me

**Comment by jfdm** [comment]
I will look at this on Thursday. 

One thing to speed up acceptance is to ensure that the commit messages are sufficient for such a big change to the testing infrastructure (i.e. having changes only specified in a GitHub PR messages is not sufficient), and also to ensure that the CHANGELOG has been accurately updated. 

Knock-on aspects of the change such as developer documentation and code commentary could be published later but not too much later.

**Comment by jfdm** [comment]
From a first glance, my gut feeling is that it will most likely be 'yes we can merge'  **but** once several amendments have been made. I won't know what the precise amendments will be until I look at it in more detail later. But most of them will be about software quality and maintenance; the 'easy part' coding is over.

**Comment by melted** [comment]
I'll look at it after work.

**Comment by jfdm** [comment]
@melted No need to look at it, I shall do it this week. I think there are blocks to this being merged straight away but want to look at the PR in detail.

**Comment by jfdm** [comment]
From a quick look at the PR, I think it can be merged once the following have been addressed:

+ [ ] The commit messages need to be self-contained and not reference a GitHub PR message. The section *Summary of Changes* from @gpyh top comment in this thread should be used in the commit message that contains the changes.
+ [ ] Information in the README.md file in tests should clearly state how to run/add/remove/disable tests.
+ [ ] The remaining sections of @gpyh top comment should be included in the README.md file in tests to provide minimal documentation on how to use the new test suite.
+ [ ] The CHANGELOG for Idris needs to be updated to reflect the change in testing
+ [ ] @gpyh Should add themselves to the CONTRIBUTERS file.

**Comment by gpyh** [comment]
Thanks for the review. I shall do this today.

**Comment by edwinb** [comment]
For this sort of thing, I need to make sure I can understand it... Although my only comment would be to please put the "How-to" comment from the PR in a README file in the test directory too. It looks like the tests themselves work more or less the same way as before, but I can also see myself forgetting to change things if adding new test families.

Anyway, thanks for doing this! It looks like it solves a lot of the irritating issues with the test script as it stands.

**Comment by david-christiansen** [comment]
Thanks for doing this! I think @jfdm's list along with the README in the test directory are what's needed.

**Comment by gpyh** [comment]
I made (most of) the requested changes.

- Commits are properly split. Hopefully the messages state everything of interest.
- The `README.md` explains everything. Please tell me if this is too wordy.
- I added myself to `CONTRIBUTORS`.

I didn't update the CHANGELOG because I didn't really know what is relevant to be put there.

I also made a new change: now we must explicitly register the tests, as opposed to scanning the directory, looking for subdirectories with the right name. This is more verbose and adds a step when creating a new test. However, specifying which tests are incompatible with the node code generator is now much more natural.

**Comment by gpyh** [comment]
A very succinct entry has been added in the `CHANGELOG`.

Travis builds should now succeed.

**Comment by gpyh** [comment]
@jfdm I've kept tweaking this. I improved some of the code and documentation.
I also removed the overriding of `ShowDetails` is the test hook.
Tests won't print in real time with cabal (so they won't in Travis either).

I have a proper fix coming but I'll submit it when this PR is merged, along with other improvements.
Until then I promise not to amend the commits.

**Comment by jfdm** [comment]
@gpyh Thanks for the hard work! 

I am happy for this to get merged, and the PR seems to be in order. If the other maintainers  (@edwinb, @david-christiansen, @melted, @melvar, @raichoo) can understand how to work with the new test suite then it is good to go.

If no one objects within 24hrs, I shall get merging.

Again thanks for the work.

**Comment by jfdm** [comment]
From reading the documentation provided a little further, Tasty does appear to generate some artefacts in the project's root directory as part of the rerun ability. 

Can this these artefacts be added to the `.gitignore` file at some point?

This might be a good chance for a secondary check to work out what/if any other artefacts generated by tasty might need to be ignored as well.

**Comment by gpyh** [comment]
Good catch! This has been added. I also took the liberty to split the lines in `.gitignore` into various categories.

The branch has been rebased onto master (which seems to confuse GitHub, but nvm). Strangely enough, this didn't trigger a Travis build.

**Comment by gpyh** [comment]
I reset the timestamp on the commits. This made GitHub much happier and triggered a Travis build.

**Comment by jfdm** [comment]
@gpyh Thanks for the hard work.

**Comment by david-christiansen** [comment]
Indeed! Thanks for sticking with it and making this great!

**Comment by gpyh** [comment]
You're welcome. I'm glad it's merged.

**Comment by ahmadsalim** [comment]
@gpyh It seems that you haven't added your changes to the CONTRIBUTORS file, at least you are not appearing in it now.

**Comment by gpyh** [comment]
I think I accidentally dropped the commit during a rebase. I will submit another PR soon, in which I'll add myself to CONTRIBUTORS.
